aproxy.go

```diff
@@ -0,0 +1,404 @@
+			return "", fmt.Errorf("failed to parse server name extension")
+		}
+
+		for !nameList.Empty() {
+			var nameType uint8
+			var serverName cryptobyte.String
+			if !nameList.ReadUint8(&nameType) ||
+				!nameList.ReadUint16LengthPrefixed(&serverName) ||
+				serverName.Empty() {
+				return "", fmt.Errorf("failed to parse server name indication extension")
+			}
+			if nameType != 0 {
+				continue
+			}
+			if len(finalServerName) != 0 {
+				return "", fmt.Errorf("multiple names of the same name_type are prohibited in server name extension")
+			}
+			finalServerName = string(serverName)
+			if strings.HasSuffix(finalServerName, ".") {
+				return "", fmt.Errorf("SNI name ends with a trailing dot")
+			}
+		}
+	}
+	return finalServerName, nil
+}
+
+func PrereadHttpHost(conn *PrereadConn) (_ string, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("failed to preread HTTP request: %w", err)
+		}
+	}()
+
+	conn.StartPreread()
+	defer conn.RestorePreread()
+	req, err := http.ReadRequest(bufio.NewReader(conn))
+	if err != nil {
+		return "", err
+	}
+	host := req.Host
+	if host == "" {
+		return "", errors.New("http request doesn't have host")
+	}
+	return host, nil
+}
+
+func DialProxy(proxy string) (net.Conn, error) {
+	proxyAddr, err := net.ResolveTCPAddr("tcp", proxy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve proxy address: %w", err)
+	}
+	conn, err := net.DialTCP("tcp", nil, proxyAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to proxy: %w", err)
+	}
+	return conn, nil
+}
+
+func DialProxyConnect(proxy string, dst string) (net.Conn, error) {
+	conn, err := DialProxy(proxy)
+	if err != nil {
+		return nil, err
+	}
+	request := http.Request{
+		Method: "CONNECT",
+		URL: &url.URL{
+			Host: dst,
+		},
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Header: map[string][]string{
+			"User-Agent": {fmt.Sprintf("aproxy/%s", version)},
+		},
+		Host: dst,
+	}
+	err = request.Write(conn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send connect request to http proxy: %w", err)
+	}
+	response, err := http.ReadResponse(bufio.NewReaderSize(conn, 0), &request)
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("proxy return %d response for connect request", response.StatusCode)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to receive http connect response from proxy: %w", err)
+	}
+	return conn, nil
+}
+
+func GetOriginalDst(conn *net.TCPConn) (*net.TCPAddr, error) {
+	file, err := conn.File()
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			slog.Error("failed to close the duplicated TCP socket file descriptor")
+		}
+	}(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert connection to file: %w", err)
+	}
+	var sockaddr [16]byte
+	size := 16
+	_, _, e := syscall.Syscall6(
+		syscall.SYS_GETSOCKOPT,
+		file.Fd(),
+		syscall.SOL_IP,
+		80, // SO_ORIGINAL_DST
```

Comment:
Is there any simpler way to get the original dest using the `net.TCPConn` directly? I guess not.

Maybe we can learn something from this function? https://sourcegraph.com/github.com/v2ray/v2ray-core/-/blob/transport/internet/tcp/sockopt_linux.go?L15, which uses `syscall` but not `unsafe`.

If we really do need to use `unsafe` here, I'd put this function in a separate file, `originaldest.go` or something, so we're limiting the scope of `unsafe` to just that one file/function.

Reply:
Also, I presume this only handles IPv4. Do we need to handle IPv6?

Reply:
I don't feel like using `syscall.GetsockoptIPv6Mreq` since it's based on the coincidence that two syscalls have the same input argument size, which can be confusing at first glance.

I think I will create a separate file. Do you think we should put it in the project root or in an internal module like `internal/syscall/syscall_linux.go`? I might need to add a few other syscalls in future versions.

As for IPv6, we only need to support IPv4 for now, but we might need to handle IPv6 in the future.